server/sonar-server/src/main/java/org/sonar/server/computation/step/QualityGateEventsStep.java

```diff
@@ -0,0 +1,133 @@
+    if (!currentStatus.isPresent()) {
+      return;
+    }
+
+    Optional<MeasureDto> previousAlertStatus = measureRepository.findPrevious(project, CoreMetrics.ALERT_STATUS);
+
+    checkQualityGateStatusChange(project, currentStatus.get(), previousAlertStatus);
+  }
+
+  private void checkQualityGateStatusChange(Component project, BatchReport.Measure currentStatus, Optional<MeasureDto> pastStatus) {
+    String alertText = currentStatus.getAlertText();
+    MetricLevel alertLevel = MetricLevel.valueOf(currentStatus.getAlertStatus());
+    if (pastStatus.isPresent()) {
+      MetricLevel pastAlertLevel = MetricLevel.valueOf(pastStatus.get().getAlertStatus());
+      if (pastAlertLevel != alertLevel) {
+        // The alert status has changed
+        String alertName = buildName(pastStatus.get(), currentStatus);
+        createEvent(project, alertName, alertText);
+        // There was already a Orange/Red alert, so this is no new alert: it has just changed
+        // boolean isNewAlert = pastAlertLevel == MetricLevel.OK;
+        // notifyUsers(project, alertName, alertText, alertLevel, isNewAlert);
+      }
+    } else if (alertLevel != MetricLevel.OK) {
+      // There were no defined alerts before, so this one is a new one
+      String alertName = buildName(currentStatus);
+      createEvent(project, alertName, alertText);
+      // notifyUsers(project, alertName, alertText, alertLevel, true);
+    }
+  }
+
+  private enum MetricLevel {
+    OK("Green"), WARN("Orange"), ERROR("Red");
+
+    private String colorName;
+
+    MetricLevel(String colorName) {
+      this.colorName = colorName;
+    }
+
+    public String getColorName() {
+      return colorName;
+    }
+  }
+
+  // protected void notifyUsers(Component project, String alertName, String alertText, MetricLevel alertLevel, boolean isNewAlert) {
+  // Notification notification = new Notification("alerts")
+  // .setDefaultMessage("Alert on " + project.getName() + ": " + alertName)
+  // .setFieldValue("projectName", project.getName())
+  // .setFieldValue("projectKey", project.getKey())
+  // .setFieldValue("projectId", String.valueOf(project.getId()))
+  // .setFieldValue("alertName", alertName)
+  // .setFieldValue("alertText", alertText)
+  // .setFieldValue("alertLevel", alertLevel.toString())
+  // .setFieldValue("isNewAlert", Boolean.toString(isNewAlert));
+  // notificationManager.scheduleForSending(notification);
+  // }
+
+  private String buildName(MeasureDto pastStatus, BatchReport.Measure currentStatus) {
+    return buildName(currentStatus) + " (was " + buildName(pastStatus) + ")";
+  }
+
+  private String buildName(BatchReport.Measure currentStatus) {
+    return MetricLevel.valueOf(currentStatus.getAlertStatus()).getColorName();
+  }
+
+  private String buildName(MeasureDto currentStatus) {
+    return MetricLevel.valueOf(currentStatus.getAlertStatus()).getColorName();
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Possible null pointer dereference in org.sonar.server.computation.step.QualityGateEventsStep.buildName(MeasureDto) due to return value of called method [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=findbugs%3ANP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE)